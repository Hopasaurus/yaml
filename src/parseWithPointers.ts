import { IParserResult, IParserResultPointers } from '@stoplight/types';
import { load as loadAST, YAMLException } from 'yaml-ast-parser';

// import { IParser, IParserResult, IParserResultPointers, MessageSeverity } from '../../types';

const get = require('lodash/get');

export interface IYamlParserOpts {
  // Increase performance by limiting how deep into the object we recurse to generate pointers.
  maxPointerDepth?: number;
}

export const parseWithPointers = <T>(value: string, opts: IYamlParserOpts = {}): IParserResult<T> => {
  const parsed: IParserResult = {
    data: {},
    pointers: {},
    validations: [],
  };

  if (!value || !value.trim().length) return parsed;

  const ast = loadAST(value);
  if (!ast) return parsed;

  const lineMap = computeLineMap(value);

  parsed.pointers = {
    '': getLoc(lineMap, {
      start: 0,
      end: ast.endPosition,
    }),
  };

  parsed.data = walk([], {}, ast.mappings, parsed.pointers, lineMap, 1, opts);

  if (ast.errors) {
    parsed.validations = transformErrors(ast.errors);
  }

  return parsed;
};

/**
 * A performant way to find the appropriate line for the given position.
 *
 * This is key to making the yaml line mapping performant.
 */
export const lineForPosition = (pos: number, lines: number[], start: number = 0, end?: number): number => {
  // position 0 is always line 1
  if (pos === 0) {
    return 1;
  }

  // start with max range, 0 - lines.length
  if (typeof end === 'undefined') {
    end = lines.length;
  }

  // target should be the halfway point between start and end
  const target = Math.floor((end - start) / 2) + start;
  if (pos >= lines[target] && !lines[target + 1]) {
    return target + 1;
  }

  // if pos is between target and the next line's position, we're good!
  const nextLinePos = lines[Math.min(target + 1, lines.length)];
  if (pos >= lines[target] && pos <= nextLinePos) {
    if (pos === nextLinePos) {
      return target + 2;
    }

    return target + 1;
  }

  // if pos is above the current line position, then we need to go "up"
  if (pos > lines[target]) {
    return lineForPosition(pos, lines, target + 1, end);
  } else {
    // else we take the bottom half
    return lineForPosition(pos, lines, start, target - 1);
  }
};

const walk = (
  path: string[],
  container: any,
  nodes: any[],
  pointers: IParserResultPointers,
  lineMap: number[],
  depth: number,
  opts: IYamlParserOpts
) => {
  if (opts.maxPointerDepth && opts.maxPointerDepth < depth) return container;

  for (const i in nodes) {
    if (!nodes.hasOwnProperty(i)) continue;

    const index = parseInt(i);
    const node = nodes[index];
    const key = node.key ? node.key.value : index;
    const nodePath = path.concat(key);

    pointers[`/${nodePath.join('/')}`] = getLoc(lineMap, {
      start: node.startPosition,
      end: node.endPosition,
    });

    const mappings = get(node, 'mappings', get(node, 'value.mappings'));
    if (mappings) {
      container[key] = walk(nodePath, {}, mappings, pointers, lineMap, depth + 1, opts);
      continue;
    }

    const items = get(node, 'items', get(node, 'value.items'));
    if (items) {
      container[key] = walk(nodePath, [], items, pointers, lineMap, depth + 1, opts);
      continue;
    }

    if (node) {
      if (node.hasOwnProperty('valueObject')) {
        container[key] = node.valueObject;
      } else if (node.hasOwnProperty('value')) {
        if (node.value && node.value.hasOwnProperty('valueObject')) {
          container[key] = node.value.valueObject;
        } else if (node.value && node.value.hasOwnProperty('value')) {
          container[key] = node.value.value;
        } else {
          container[key] = node.value;
        }
      }
    } else {
      container[key] = node;
    }
  }

  return container;
};

// builds up the line map, for use by linesForPosition
const computeLineMap = (input: string) => {
  const lines = input.split(/\n/);
  const lineMap: number[] = [];

  let sum = 0;
  for (const line of lines) {
    lineMap.push(sum);
    sum += line.length + 1;
  }

  return lineMap;
};

const getLoc = (lineMap: number[], { start = 0, end = 0 }) => {
  return {
    start: { line: lineForPosition(start, lineMap) },
    end: { line: lineForPosition(end, lineMap) },
  };
};

const transformErrors = (errors: YAMLException[]): any[] => {
  const validations: any[] = [];
  for (const error of errors) {
    const validation: any = {
      ruleId: error.name,
      msg: error.reason,
      // FIXME when LogLevel enum published in types package: error.isWarning ? LogLevel.Warn : LogLevel.Fatal
      level: error.isWarning ? 40 : 60,
    };

    if (error.mark && error.mark.line) {
      validation.location = {
        start: {
          line: error.mark.line,
        },
      };
    }

    validations.push(validation);
  }

  return validations;
};
