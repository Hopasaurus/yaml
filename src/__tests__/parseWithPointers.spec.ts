import { lineForPosition, yamlParser } from '../parseWithPointers';
import * as HugeJSON from './fixtures/huge-json.json';
import { HugeYAML } from './fixtures/huge-yaml';

const diverse = `---
  # <- yaml supports comments, json does not
  # did you know you can embed json in yaml?
  # try uncommenting the next line
  # { foo: 'bar' }

  json:
    - rigid
    - better for data interchange
  yaml:
    - slim and flexible
    - 123
    - true
  object:
    key: value
    array:
      - null_value:
      - boolean: true
      - integer: 1
  paragraph: >
      Blank lines denote

      paragraph breaks
  content: |-
      Or we
      can auto
      convert line breaks
      to save space`;

describe('yaml parser', () => {
  test('lineForPosition simple', () => {
    const lines = [0, 13, 22, 36];
    // first line
    expect(lineForPosition(0, lines)).toEqual(1);

    expect(lineForPosition(12, lines)).toEqual(1);
    expect(lineForPosition(35, lines)).toEqual(3);

    // last line
    expect(lineForPosition(36, lines)).toEqual(4);
    expect(lineForPosition(39, lines)).toEqual(4);
  });

  test('lineForPosition diverse', () => {
    const lines = [
      0,
      4,
      49,
      94,
      129,
      148,
      149,
      157,
      169,
      203,
      211,
      235,
      245,
      256,
      266,
      281,
      292,
      312,
      334,
      353,
      368,
      393,
      394,
      417,
      431,
      443,
      484,
    ];

    // first line
    expect(lineForPosition(0, lines)).toEqual(1);
    expect(lineForPosition(1, lines)).toEqual(1);
    expect(lineForPosition(4, lines)).toEqual(2);
    expect(lineForPosition(5, lines)).toEqual(2);
    expect(lineForPosition(51, lines)).toEqual(3);
    expect(lineForPosition(255, lines)).toEqual(13);
    expect(lineForPosition(256, lines)).toEqual(14);
    expect(lineForPosition(257, lines)).toEqual(14);
    expect(lineForPosition(417, lines)).toEqual(24);
    expect(lineForPosition(418, lines)).toEqual(24);
    expect(lineForPosition(483, lines)).toEqual(26);
    expect(lineForPosition(484, lines)).toEqual(27);
    expect(lineForPosition(599, lines)).toEqual(27);
  });

  test('parse simple', () => {
    expect(
      yamlParser()(`hello: world
address:
  street: 123`)
    ).toMatchSnapshot();
  });

  test('parse diverse', () => {
    const result = yamlParser()(diverse);
    delete result.validations;
    expect(result).toMatchSnapshot();
  });

  test('parse huge', () => {
    const result = yamlParser()(HugeYAML);
    expect(result.data).toEqual(HugeJSON);
  });

  test('max depth option', () => {
    const result = yamlParser({ maxPointerDepth: 3 })(`prop1: true
prop2: true
prop3:
  prop3-1:
    prop3-1-1:
      prop3-1-1-1: true
    prop3-1-2:
      - one
      - two`);

    expect(result.pointers).toMatchSnapshot();
  });

  test('report errors', () => {
    const result = yamlParser({ maxPointerDepth: 3 })(`prop1: true
prop2: true
  inner 1
  val: 2`);

    expect(result.validations).toEqual([
      {
        level: 60,
        location: { start: { line: 3 } },
        msg: 'bad indentation of a mapping entry',
        ruleId: 'YAMLException',
      },
      {
        level: 60,
        location: { start: { line: 3 } },
        msg: 'incomplete explicit mapping pair; a key node is missed',
        ruleId: 'YAMLException',
      },
    ]);
  });
});
