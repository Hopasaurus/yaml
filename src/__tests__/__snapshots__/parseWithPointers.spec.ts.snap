// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml parser max depth option 1`] = `
Object {
  "": Object {
    "end": Object {
      "line": 9,
    },
    "start": Object {
      "line": 1,
    },
  },
  "/prop1": Object {
    "end": Object {
      "line": 1,
    },
    "start": Object {
      "line": 1,
    },
  },
  "/prop2": Object {
    "end": Object {
      "line": 2,
    },
    "start": Object {
      "line": 2,
    },
  },
  "/prop3": Object {
    "end": Object {
      "line": 9,
    },
    "start": Object {
      "line": 3,
    },
  },
  "/prop3/prop3-1": Object {
    "end": Object {
      "line": 9,
    },
    "start": Object {
      "line": 4,
    },
  },
  "/prop3/prop3-1/prop3-1-1": Object {
    "end": Object {
      "line": 6,
    },
    "start": Object {
      "line": 5,
    },
  },
  "/prop3/prop3-1/prop3-1-2": Object {
    "end": Object {
      "line": 9,
    },
    "start": Object {
      "line": 7,
    },
  },
}
`;

exports[`yaml parser parse diverse 1`] = `
Object {
  "data": Object {
    "content": "Or we
can auto
convert line breaks
to save space",
    "json": Array [
      "rigid",
      "better for data interchange",
    ],
    "object": Object {
      "array": Array [
        Object {
          "null_value": null,
        },
        Object {
          "boolean": true,
        },
        Object {
          "integer": 1,
        },
      ],
      "key": "value",
    },
    "paragraph": "Blank lines denote
paragraph breaks
",
    "yaml": Array [
      "slim and flexible",
      123,
      true,
    ],
  },
  "pointers": Object {
    "": Object {
      "end": Object {
        "line": 28,
      },
      "start": Object {
        "line": 1,
      },
    },
    "/content": Object {
      "end": Object {
        "line": 28,
      },
      "start": Object {
        "line": 24,
      },
    },
    "/json": Object {
      "end": Object {
        "line": 10,
      },
      "start": Object {
        "line": 7,
      },
    },
    "/json/0": Object {
      "end": Object {
        "line": 8,
      },
      "start": Object {
        "line": 8,
      },
    },
    "/json/1": Object {
      "end": Object {
        "line": 9,
      },
      "start": Object {
        "line": 9,
      },
    },
    "/object": Object {
      "end": Object {
        "line": 20,
      },
      "start": Object {
        "line": 14,
      },
    },
    "/object/array": Object {
      "end": Object {
        "line": 20,
      },
      "start": Object {
        "line": 16,
      },
    },
    "/object/array/0": Object {
      "end": Object {
        "line": 17,
      },
      "start": Object {
        "line": 17,
      },
    },
    "/object/array/0/null_value": Object {
      "end": Object {
        "line": 17,
      },
      "start": Object {
        "line": 17,
      },
    },
    "/object/array/1": Object {
      "end": Object {
        "line": 18,
      },
      "start": Object {
        "line": 18,
      },
    },
    "/object/array/1/boolean": Object {
      "end": Object {
        "line": 18,
      },
      "start": Object {
        "line": 18,
      },
    },
    "/object/array/2": Object {
      "end": Object {
        "line": 19,
      },
      "start": Object {
        "line": 19,
      },
    },
    "/object/array/2/integer": Object {
      "end": Object {
        "line": 19,
      },
      "start": Object {
        "line": 19,
      },
    },
    "/object/key": Object {
      "end": Object {
        "line": 15,
      },
      "start": Object {
        "line": 15,
      },
    },
    "/paragraph": Object {
      "end": Object {
        "line": 23,
      },
      "start": Object {
        "line": 20,
      },
    },
    "/yaml": Object {
      "end": Object {
        "line": 14,
      },
      "start": Object {
        "line": 10,
      },
    },
    "/yaml/0": Object {
      "end": Object {
        "line": 11,
      },
      "start": Object {
        "line": 11,
      },
    },
    "/yaml/1": Object {
      "end": Object {
        "line": 12,
      },
      "start": Object {
        "line": 12,
      },
    },
    "/yaml/2": Object {
      "end": Object {
        "line": 13,
      },
      "start": Object {
        "line": 13,
      },
    },
  },
}
`;

exports[`yaml parser parse simple 1`] = `
Object {
  "data": Object {
    "address": Object {
      "street": 123,
    },
    "hello": "world",
  },
  "pointers": Object {
    "": Object {
      "end": Object {
        "line": 3,
      },
      "start": Object {
        "line": 1,
      },
    },
    "/address": Object {
      "end": Object {
        "line": 3,
      },
      "start": Object {
        "line": 2,
      },
    },
    "/address/street": Object {
      "end": Object {
        "line": 3,
      },
      "start": Object {
        "line": 3,
      },
    },
    "/hello": Object {
      "end": Object {
        "line": 1,
      },
      "start": Object {
        "line": 1,
      },
    },
  },
  "validations": Array [],
}
`;
